SceneRefAttributePropertyDrawer.cs

```diff
@@ -14,16 +17,106 @@ namespace KBCore.Refs
+        string typeName;
+
+        SceneRefAttribute sceneRefAttribute => (SceneRefAttribute)attribute;
+
         public override void OnGUI(Rect position, SerializedProperty property, GUIContent label)
         {
+            if (!isInitialized)
+            {
+                Initialize(property);
+            }
+
+            if (isValidFieldType)
+            {
+                bool isSatisfied = IsSatisfied(property);
+                if (!isSatisfied)
+                {
+                    Rect helpBoxPos = position;
+                    helpBoxPos.height = EditorGUIUtility.singleLineHeight * 2;
+                    string message = $"{property.propertyPath} missing {typeName} reference on {sceneRefAttribute.Loc}!";
+                    EditorGUI.HelpBox(helpBoxPos, message, MessageType.Error);
+                    position.height = EditorGUI.GetPropertyHeight(property, label);
+                    position.y += helpBoxPos.height;
+                }
+            }
+
             bool wasEnabled = GUI.enabled;
             GUI.enabled = false;
             EditorGUI.PropertyField(position, property, label, true);
             GUI.enabled = wasEnabled;
         }
 
-        public override float GetPropertyHeight(SerializedProperty property, GUIContent label) 
-            => EditorGUI.GetPropertyHeight(property, label);
+        void Initialize(SerializedProperty property)
+        {
+            isInitialized = true;
+
+            // the type wont change, so we only need to initialize these values once
+
+            elementType = fieldInfo.FieldType;
+            if (typeof(ISerializableRef).IsAssignableFrom(elementType))
+            {
+                var interfaceType = elementType.GetInterfaces().FirstOrDefault(type =>
+                    type.IsGenericType && type.GetGenericTypeDefinition() == typeof(ISerializableRef<>));
+                if (interfaceType != null)
+                {
+                    elementType = interfaceType.GetGenericArguments()[0];
+                }
+            }
+
+            isValidFieldType = typeof(Component).IsAssignableFrom(elementType)
+                && (property.propertyType == SerializedPropertyType.ObjectReference);
+
+            typeName = fieldInfo.FieldType.Name;
+            if (fieldInfo.FieldType.IsGenericType && fieldInfo.FieldType.GenericTypeArguments.Length >= 1)
+            {
+                typeName = typeName.Replace("`1", $"<{fieldInfo.FieldType.GenericTypeArguments[0].Name}>");
+            }
+        }
+
+        /// <summary>Is this field Satisfied with a value or optional</summary>
+        private bool IsSatisfied(SerializedProperty property)
+        {
+            if (sceneRefAttribute.HasFlags(Flag.Optional))
+            {
+                return true;
+            }
+            return CheckValue(property);
+        }
+
+        private bool CheckValue(SerializedProperty property)
+        {
+            bool hasValue = false;
+            if (property.isArray)
```

Comment:
This branch can be removed right? It looks like `CheckValue` will never be invoked for array/list types anyways, since `isValidFieldType` will only be true for component refs

Reply:
oh yeah, the property will never be an array in this case so it can be removed